src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -531,6 +534,52 @@ protected void AddProjectReference(ProjectReference projectReference)
             }
         }
 
+        protected bool CanAddProjectReference(ProjectReference projectReference)
+        {
+            if (projectReference.ProjectId == this.Id)
+            {
+                // cannot self reference
+                return false;
+            }
+
+            if (_projectReferences.Contains(projectReference))
+            {
+                // already have this reference
+                return false;
+            }
+
+            var project = this.ProjectTracker.GetProject(projectReference.ProjectId);
+            if (project != null)
+            {
+                // cannot add a reference to a project that references us (it would make a cycle)
+                return !project.TransitivelyReferences(this.Id);
+            }
+
+            return true;
+        }
+
+        private bool TransitivelyReferences(ProjectId projectId)
```

Comment:
I'm worried about the potential exponential runtime growth of this. Since this is only run when references are added (infrequently) I assume we can afford an allocation to track the visited nodes, right?